prow/pluginhelp/hook/hook.go

```diff
@@ -0,0 +1,366 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/pluginhelp/externalplugins"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+// TODO: unit test to ensure that external plugins with the same name have the same endpoint and events.
+
+const (
+	// newRepoDetectionLimit is the maximum allowable time before a repo will appear in the help
+	// information if it is a new repo that is only referenced via its parent org in the config.
+	// (i.e. max time before help is available for brand new "kubernetes/foo" repo if only
+	// "kubernetes" is listed in the config)
+	newRepoDetectionLimit = time.Hour
+)
+
+type pluginAgent interface {
+	Config() *plugins.Configuration
+}
+
+type githubClient interface {
+	GetRepos(org string, isUser bool) ([]github.Repo, error)
+}
+
+type HelpAgent struct {
+	log *logrus.Entry
+	pa  pluginAgent
+	oa  *orgAgent
+}
+
+func NewHelpAgent(pa pluginAgent, ghc githubClient) *HelpAgent {
+	l := logrus.WithField("client", "plugin-help")
+	return &HelpAgent{
+		log: l,
+		pa:  pa,
+		oa:  newOrgAgent(l, ghc, newRepoDetectionLimit),
+	}
+}
+
+func (ha *HelpAgent) generateNormalPluginHelp(config *plugins.Configuration, revMap map[string][]string) (allPlugins []string, pluginHelp map[string]pluginhelp.PluginHelp) {
+	pluginHelp = map[string]pluginhelp.PluginHelp{}
+	for name, provider := range plugins.HelpProviders() {
+		allPlugins = append(allPlugins, name)
+		if provider == nil {
+			ha.log.Warnf("No help is provided for plugin '%s'.", name)
+			continue
+		}
+		help, err := provider(config, revMap[name])
+		if err != nil {
+			ha.log.WithError(err).Errorf("Generating help from normal plugin '%s'.", name)
+			continue
+		}
+		help.Events = plugins.EventsForPlugin(name)
+		pluginHelp[name] = *help
+	}
+	return
+}
+
+func (ha *HelpAgent) generateExternalPluginHelp(config *plugins.Configuration, revMap map[string][]string) (allPlugins []string, pluginHelp map[string]pluginhelp.PluginHelp) {
+	externals := map[string]plugins.ExternalPlugin{}
+	for _, exts := range config.ExternalPlugins {
+		for _, ext := range exts {
+			externals[ext.Name] = ext
+		}
+	}
+
+	type externalResult struct {
+		name string
+		help *pluginhelp.PluginHelp
+	}
+	externalResultChan := make(chan externalResult, len(externals))
+	for _, ext := range externals {
+		allPlugins = append(allPlugins, ext.Name)
+		go func(ext plugins.ExternalPlugin) {
+			help, err := externalHelpProvider(ha.log, ext.Endpoint)(revMap[ext.Name])
+			if err != nil {
+				ha.log.WithError(err).Errorf("Getting help from external plugin '%s'.", ext.Name)
+				help = nil
+			} else {
+				help.Events = ext.Events
+			}
+			externalResultChan <- externalResult{name: ext.Name, help: help}
+		}(ext)
+	}
+
+	pluginHelp = map[string]pluginhelp.PluginHelp{}
+	timeout := time.After(time.Second)
```

Comment:
Why 1s?

Reply:
Because the help information is generated as the result of a browser request, so I didn't think it should be allowed to hang for a long time. Also this is an in-cluster network call so it should succeed quickly.
Is this reasonable?

Reply:
Not sure. Keep in mind that external plugins may not run in the same cluster as hook. Definitely a thing to monitor once this is merged.